Stratis.Bitcoin/Utilities/AsyncLock.cs

```diff
@@ -1,131 +1,47 @@
-    /// only guarantees to actually run one task at the time, but if the task awaits, it is not considered 
-    /// as running and another task can be scheduled and run instead within the context of the exclusive 
-    /// scheduler. This means that the tasks run within both exclusive and concurrent schedulers 
-    /// must not await, otherwise there is a risk of a race condition. Thus in order to use this locking 
-    /// mechanism, one needs to first break up the asynchronous code to synchronous pieces and only then 
-    /// schedule the synchronous parts.
-    /// </para>
-    /// </remarks>
-    public class AsyncLock : IAsyncLock
-    {
-        /// <inheritdoc />
-        private CancellationTokenSource cancellation;
-
-        /// <summary>Task factory that runs tasks using the concurrent scheduler. Serves as a reader lock.</summary>
-        private readonly TaskFactory concurrentFactory;
-
-        /// <summary>Task factory that runs tasks using the exclusive scheduler. Serves as a writer lock.</summary>
-        private readonly TaskFactory exclusiveFactory;
-
-        /// <summary>
-        /// Initializes a new instance of the object with ability to cancel locked tasks. 
-        /// </summary>
-        /// <param name="cancellation">Cancellation source to allow cancel the tasks run by the schedulers.</param>
-        /// <param name="maxItemsPerTask">Number of exclusive tasks to process before checking concurrent tasks.</param>
-        public AsyncLock(CancellationTokenSource cancellation = null, int maxItemsPerTask = 5)
+        public AsyncLock()
         {
-            this.cancellation = cancellation ?? new CancellationTokenSource();
-            int defaultMaxConcurrencyLevel = Environment.ProcessorCount; 
-            int defaultMaxItemsPerTask = maxItemsPerTask; 
-            var schedulerPair = new ConcurrentExclusiveSchedulerPair(TaskScheduler.Default, defaultMaxConcurrencyLevel, defaultMaxItemsPerTask);
-            this.concurrentFactory = new TaskFactory(schedulerPair.ConcurrentScheduler);
-            this.exclusiveFactory = new TaskFactory(schedulerPair.ExclusiveScheduler);
+            this.semaphore = new SemaphoreSlim(1, 1);
+            this.releaser = Task.FromResult((IDisposable)new Releaser(this));
         }
 
-        /// <inheritdoc />
-        public Task ReadAsync(Action func)
+        public Task<IDisposable> LockAsync()
         {
-            return this.concurrentFactory.StartNew(func, this.cancellation.Token);
-        }
+            var wait = this.semaphore.WaitAsync();
 
-        /// <inheritdoc />
-        /// <remarks>See warning in <see cref="AsyncLock"/> remarks section.</remarks>
-        public Task<T> ReadAsync<T>(Func<T> func)
-        {
-            return this.concurrentFactory.StartNew(func, this.cancellation.Token);
+            return wait.IsCompleted ?
+                        this.releaser :
+                        wait.ContinueWith((_, state) => (IDisposable)state,
```

Comment:
This can be written nicer

Reply:
I would raise a question - Do we not want LockAsync method to accept cancellation token? Maybe optional one? Would that be beneficial to any use case?

Reply:
Indeed we want.
Do we have to specify a `TaskScheduler.Default` ?